compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -2602,6 +2611,15 @@ pub enum Rvalue<'tcx> {
     /// initialized but its content as uninitialized. Like other pointer casts, this in general
     /// affects alias analysis.
     ShallowInitBox(Operand<'tcx>, Ty<'tcx>),
+
+    /// A virtual ref is equivalent to a read from a reference/pointer at the
+    /// codegen level, but is treated specially by drop elaboration. When such a read happens, it
+    /// is guaranteed that the only use of the returned value is a deref operation, immediately
+    /// followed by one or more projections. Drop elaboration treats this virtual ref as if the
+    /// read never happened and just projects further. This allows simplifying various MIR
+    /// optimizations and codegen backends that previously had to handle deref operations anywhere
+    /// in a place.
```

Comment:
I am confused by this comment.

First of all, it seems to read from a place, not a reference/pointer, so the first sentence doesn't entirely parse for me.

Secondly, "the only use of the returned value is a deref operation" -- does this mean that the place must have pointer/reference type, so that a deref projection can be applied to it? That would probably be worth enforcing in some places.

And finally, who is making this guarantee? Is it UB to violate it? Or is there some syntactic invariant which ensures the guarantee to hold on all MIR programs?

Also Cc @JakobDegen who did a lot of work on the MIR semantics docs recently.

Reply:
Yeah, agreed. Whatever comment goes here needs to describe 1) what the operational semantics of this `Rvalue` are and 2) what the additional restrictions (if any) for its existence are. If the way drop elaboration treats this `Rvalue` is not obvious from that description, then that is something that should be noted here also, but that also needs to be fully described in the docs for drop elaboration (probably that is on `TerminatorKind::Drop` or `MirPhase::DropsElaborated`)

Reply:
It might seem unclear what's the purpose of this operation, let me try to briefly explain what it does so my comments below make more sense.
 We created this operation because trying to move `deref_separator` before `elaborate_drops` caused cryptic bugs to appear along with the code generation, such as `elaborate_drop` trying to drop temp values created by `MirPatch` or treating the MIR body completely differently (`elaborate_replace` trying to replace terminator of the temp value) hence the cryptic crash that appeared when building MIRI.

 So with @oli-obk suggestion, we created this new `Rvalue` (which is almost entirely the same as `Rvalue::Use::Operand::Copy()`), to see which statements are the results of `deref_separator` that allow us to undo changes made by `deref_separator` for the duration of `elaborate_drops` then redo it again at the end of `elaborate_drops` again. 

> First of all, it seems to read from a place, not a reference/pointer, so the first sentence doesn't entirely parse for me.

It reads from a `Place` which has `Deref` in it i.e 

```rust
 switchInt((*(*(*(*_4))))) 
```

> Secondly, "the only use of the returned value is a deref operation" -- does this mean that the place must have pointer/reference type, so that a deref projection can be applied to it? That would probably be worth enforcing in some places.

Deref projection is already applied to that place. Guarantee is enforced by `transform/validate` and `deref_separator` (that's the only place `VirtualRef` is used)

Reply:
> It reads from a Place which has Deref in it i.e

So in a `VirtualRef(Place<'tcx>)`, the `Place` must have a `Deref` in it?

Also your example doesn't contain a `VirtualRef` so it doesn't help me understand.

> Deref projection is already applied to that place. 

You cannot deref a place, you can only deref a value (of pointer type).
So to deref a place, you first have to `Operand::Copy` that place to get a value, and then deref it.
(Unfortunately MIR leaves this implicit, which often confuses people. But when talking about MIR semantics we should be clear about what happens.)

> this new Rvalue (which is almost entirely the same as Rvalue::Use::Operand::Copy())

That is a good way to summarize its semantics. :D  But then its name makes very little sense. If it does a copy of a place why does it have "ref" in its name?

Reply:
Talking about MIR is confusing 😅 

Yeah now that I take another look at it, you are right the name is indeed confusing. It should be something like 
`Rvalue::DerefCopy` since the only thing it does is copy each `Deref` to temp values i.e example below
```rust
          StorageLive(_12);                // scope 1 at $DIR/derefer_complex_case.rs:4:10: 4:13
-         _12 = (*((_7 as Some).0: &i32)); // scope 1 at $DIR/derefer_complex_case.rs:4:10: 4:13
+         StorageLive(_15);                // scope 1 at $DIR/derefer_complex_case.rs:4:10: 4:13
+         _15 = virt ((_7 as Some).0: &i32); // scope 1 at $DIR/derefer_complex_case.rs:4:10: 4:13  
//         ^ This is the VirtualRef value as you can see it's only used for copying the deref'ed value which is derefered inside deref_separator 
+         _12 = (*_15);                    // scope 1 at $DIR/derefer_complex_case.rs:4:10: 4:13
+         StorageDead(_15);  
```

Reply:
It's basically `Copy` so maybe it should be named that. Maybe `CopyForDeref` or so -- the idea is that in your example, the only allowed use of `_15` is via `*_15`, right? And I guess it also cannot be `&*_15` as that wouldn't actually do a load/copy?

This example *does* answer one of my other questions: yes the place operand of `virt` needs to be of pointer type. It's a place containing a pointer.

I don't like the "virtual" in its name because *the load actually happens*, there's nothing "virtual" about it.

Reply:
> the idea is that in your example, the only allowed use of _15 is via *_15, right? And I guess it also cannot be &*_15 as that wouldn't actually do a load/copy?

Yes **only** *_15 is allowed.
I'm bad at naming things... I like `CopyForDeref` name its concise.

Reply:
Don't worry, naming things is one of the [hardest problems in CS](https://www.karlton.org/2017/12/naming-things-hard/). :)

Reply:
>  Guarantee is enforced by transform/validate

I don't see that in the diff, am I looking in the wrong spot?
This seems somewhat expensive to check for.

Reply:
I'm going to try and put up an MCP today that proposes a proper distinction between analysis and optimization MIR. I'd want to evaluate how much sense this change makes in the context of a proper distinction in that space. In other words, I may be philosophizing too much (again) about the "right" solution to these things, but I'm honestly unconvinced that this complexity is worth it

Reply:
It's a somewhat implicit guarantee. The way`deref_separator` works guarantee it. https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir_transform/src/deref_separator.rs

Reply:
> I'm honestly unconvinced that this complexity is worth it
The plan is to remove `Deref` from projections entirely hopefully when that happens most of these complexities will be _less_ complex. Until then all we could do is incrementally take small steps toward that goal.

Reply:
Yeah, but this new `Rvalue` just seems like a workaround for the problems created by attempting to remove the `Deref` projections in analysis MIR. I don't see how this will go away. Maybe the other simplifications will outweigh this one, but I'm not at all convinced of that.